### ISSUE TITLE
RTK to normal GPS handover

### ISSUE THREAD

**Issue by thomasgubler** [PX4/Firmware#9431]
Goal: verify and then improve and robustify PX4 capabilities with RTK

1.  asses current state and cornercases of RTK support in the estimator
  1. what happens if connection to reference is lost and variance of measurement grows suddenly
  1. what happens if RTK is regained
  1. etc.
1. Discuss required modifications and scope
1. implement robustifications and improvements found in 1)

**Comment by thomasgubler** [comment]
Blocked by: https://github.com/PX4/Firmware/issues/9422 and https://github.com/PX4/Firmware/issues/9432

**Comment by JohnSnowball** [comment]
I guess I might give you a little help. I'm currently using an Ashtech RTK + ublox m8n. At the begining, I use only RTK, but many times, RTK lost signal in air. Now I'm using dual-gps, and I can observe the switch between the two gps sometimes. During switch, from the plane's behaviour, it's smooth, however, eph would raise rapidly. 
Then, when RTCM signal is lost(bad environment), RTK fix type will drop to SBAS or 3DLock depends on different brands. In this case, ublox m8n could have better accuracy and stability!

**Comment by thomasgubler** [comment]
➤ Paul Riseborough commented:

Blocking issue https://github.com/PX4/Firmware/issues/9422 has been resolved. A branch to test logging of a second reciever as a truth reference has been created https://github.com/priseborough/Firmware/tree/ekfGroundTruth-wip

**Comment by beat** [comment]
@thomasgubler Please use @bkueng for Beat Küng and not @beat ! Thanks :smile:

**Comment by thomasgubler** [comment]
➤ Paul Riseborough commented:

Data from a second GPs can now be recorded. With the https://github.com/priseborough/Firmware/tree/ekfGroundTruth-wip branch I was alos trying to log a groudn truth message with the equivalent loval position to enable overlay plotting with the EKF, but for a reason I have been unable to ascertain, the ground truth message is not being logged despite appearing if I run the listener command at the terminal.

**Comment by dagar** [comment]
@thomasgubler https://github.com/PX4/Firmware/blob/master/src/modules/logger/logger.cpp#L663

**Comment by priseborough** [comment]
OK, I had avoided that previously due to the inefficiency of not requiring half the data fields, but the data rate is low (5Hz) so the efficiency doesn't matter in this instance. I gave it a try here https://github.com/priseborough/Firmware/tree/ekfGroundTruth-wip-reworked , but still have the same problem - The uORB data can be read using listener, but does not show when plotting the log using the FlightPlot utility. Data for the second GPS has been logged. I must be missing something about how to add topics to the logger.

The logged data is here: https://logs.px4.io/plot_app?log=a21360e4-90b7-4032-80a2-7607fa6d3b43

**Comment by bkueng** [comment]
@priseborough the log contains the message.
`ulog_info <file.ulg>`:
```
Name (multi id, message size in bytes)    number of data points, total bytes
 actuator_controls_0 (0, 48)                  979      46992
 actuator_outputs (0, 76)                     987      75012
 battery_status (0, 63)                       201      12663
 cpuload (0, 16)                              100       1600
 ekf2_innovations (0, 164)                    474      77736
 ekf2_timestamps (0, 24)                    25058     601392
 estimator_status (0, 272)                    495     134640
 home_position (0, 47)                          2         94
 manual_control_setpoint (0, 63)               94       5922
 mission_result (0, 31)                         5        155
 ping (0, 30)                                  92       2760
 position_setpoint_triplet (0, 320)             4       1280
 rate_ctrl_status (0, 36)                    3142     113112
 safety (0, 12)                               985      11820
 sensor_combined (0, 44)                    25059    1102596
 sensor_preflight (0, 20)                     496       9920
 system_power (0, 23)                         201       4623
 telemetry_status (0, 36)                      93       3348
 vehicle_air_data (0, 24)                    7038     168912
 vehicle_attitude (0, 53)                    3030     160590
 vehicle_attitude_setpoint (0, 55)            975      53625
 vehicle_global_position (0, 64)              356      22784
 vehicle_gps_position (0, 91)                 485      44135
 vehicle_gps_position (1, 91)                 484      44044
 vehicle_land_detected (0, 16)                 98       1568
 vehicle_local_position (0, 144)              943     135792
 vehicle_local_position_groundtruth (0, 144)     484      69696
 vehicle_local_position_setpoint (0, 64)      697      44608
 vehicle_magnetometer (0, 20)                4684      93680
 vehicle_rates_setpoint (0, 24)              3142      75408
 vehicle_status (0, 39)                       103       4017
 vehicle_status_flags (0, 34)                 103       3502
```

But you forgot to set the `timestamp`.

**Comment by AuterionWrikeBot** [comment]
➤ Paul Riseborough commented:

I'm going through some solution options at the moment for GPS switching and/or blending and I think we need to move the parameters used to specify the phase centre of the GPS antenna in body frame the body frame out of the ekf2 module and into the sensors module. That enables us to specify different locations for each antenna which are published as part of the GPS message.

**Comment by bkueng** [comment]
> into the sensors module

GPS is not handled in the sensors module but comes directly from the GPS driver. Moving the params there is one option and sounds good. The other is that the GPS driver guarantees that the main GPS is on uORB instance 0 whereas the secondary is on instance 1, which can be used in ekf2 to apply the parameters.

**Comment by AuterionWrikeBot** [comment]
➤ Paul Riseborough commented:

Agree - having a guaranteed mapping from instance to physical receiver connection is essential

**Comment by AuterionWrikeBot** [comment]
➤ Paul Riseborough commented:

Created supporting functions for calculating weighting and the blended solution here: https://github.com/priseborough/Firmware/tree/pr-ekfDualGps.

The next steps are to write the code that calls these functions only when blending is required and modification of the parameters to support different body frame offsets for each receiver.

**Comment by dagar** [comment]
One idea to consider is that the actual GPS blending code (+ helpers) could live in a separate library (src/lib/gps) and called from EKF2. This would allow others to use it without the overhead and latency of a separate task consuming all gps subscriptions, blending, and republishing.

**Comment by AuterionWrikeBot** [comment]
➤ Paul Riseborough commented:

Totally agree. At the moment this is the fastest way for me to develop and test, but once I am happy with the algorithms performance, i will be asking for assistance with integrating it in a more general way as you suggest.

**Comment by priseborough** [comment]
I've finished coding the algorithms (see here) and have completed some bench and replay testing that shows them working as expected. The next step is to port them to a better location, and add the location parameters for the second receiver.

 @dagar I will be in contact to discuss porting options.

**Comment by AuterionWrikeBot** [comment]
➤ Wrike Bot commented:

@Paul Riseborough All subtasks have been completed or deferred, you might want to complete this task as well.
